### PR TITLE
Goon-style lighting

### PR DESCRIPTION
**Pull request by MrPerson**
Still WIP. Performance is not atrocious but pretty poor. Needs optimization which is why I'm making this PR. I need help with that.

I could really use a better sprite for the lights, the current one is kinda bad. I'll do one eventually but I promise it's not hard. If anybody wants to see the tool I wrote to help with this, just ask.

.I changed lights to track based on corners instead of the whole tile. That's literally how the light renders now. Additionally the sprite (kinda) transitions between the brightness in a (somewhat) smooth way.

Ask questions if you want. Obviously requires BYOND v509. Here's some pictures because it would be rude to not include some.

### CODE REVIEW

**Comment by Bawhoppen** [comment]
eh, I don't like it
also goon does this?

**Comment by WJohn** [comment]
While smooth lighting is cool, this looks far, far too blocky, and I'm not at all a fan of how it's split in the middle of walls.

**Comment by Fox-McCloud** [comment]
With a little work, I think this will eventually wind up pretty damn good.

Though, I do strongly recommend keeping absolute darkness--even in rooms with literally ZERO light, on Goon, it's still bright enough for you to be able to see literally everything. Sure, the details might be harder to make out on things, but you'd literally have no trouble making anything out.

**Comment by Bawhoppen** [comment]
Not long ago we had the ability to see everything in a room even completely dark if you just looked at your monitor closely or brought up your gamma
Now you can still hover over with your mouse and use Alt-click to see everything though

**Comment by MrStonedOne** [comment]
@Bawhoppen i fixed that when i tweaked how lighting fades off.

**Comment by Bawhoppen** [comment]
the first thing or the second thing?

File: code/modules/lighting/lighting_system.dm [hunk 1]
```diff
@@ -316,31 +382,41 @@
 	if(config.starlight)
 		update_starlight()
 
+#define MEAN4(val1, val2, val3, val4) ((val1+val2+val3+val4)/4)
```
Comment (RemieRichards) [inline code/modules/lighting/lighting_system.dm]: *0.25 instead of /4

**Comment by Incoming5643** [comment]
If full dark areas can still be looked into even if there's absolutely no sources of light in the area that's kind of a problem. I know in the goonstation version of this there's no such thing as tiles "hidden by darkness". I jokingly called it "RP Darkness" beforce, but it's not entirely a joke. Shadowlings would be in big trouble.

**Comment by Bawhoppen** [comment]
well if we're keeping full darkness we need to fix the fact you can hover over/alt-click/rightclick to see everything
but then again usually our lighting system doesn't work anyways

**Comment by MrStonedOne** [comment]
@Bawhoppen that SHOULD be fixed, for humans at least. those tiles are treated as tiles hidden by a dense object in the way.

If it isn't, i can look at finding the source of that.

**Comment by Bawhoppen** [comment]
Well ingame right now it isn't (unless I'm a lemming)

File: code/modules/lighting/lighting_system.dm [hunk 2]
```diff
@@ -72,7 +73,9 @@
 		return
 
 	radius = max(LIGHTING_MIN_RADIUS, new_luminosity)
+	radius = owner.get_light_range(radius)
```
Comment (tkdrg) [inline code/modules/lighting/lighting_system.dm]: Well you could save on var access overhead with `radius = owner.get_light_range(max(LIGHTING_MIN_RADIUS, new_luminosity))` but I'm not sure it matters a lot.

File: code/modules/lighting/lighting_system.dm [hunk 3]
```diff
@@ -22,18 +24,16 @@
...
 		Colored lights
+		Normalize lumcounts to be from 0 to 1 instead of 0 to 10 so LIGHTING_CALC is unnecessary
 */
 
 #define LIGHTING_CIRCULAR 1									//Comment this out to use old square lighting effects.
 #define LIGHTING_LAYER 15									//Drawing layer for lighting
-#define LIGHTING_CAP 10										//The lumcount level at which alpha is 0 and we're fully lit.
-#define LIGHTING_CAP_FRAC (255/LIGHTING_CAP)				//A precal'd variable we'll use in turf/redraw_lighting()
+#define LIGHTING_CAP 10										//The lumcount level at which we're fully lit.
+#define LIGHTING_CALC(value) (value/LIGHTING_CAP)
```
Comment (tkdrg) [inline code/modules/lighting/lighting_system.dm]: You can do what remie said here too, `#define LIGHTING_CALC_CONST (1/LIGHTING_CAP)` (BYOND should calculate this constant at compile-time hopefully) and then `value*LIGHTING_CALC_CONST`.

File: code/modules/lighting/lighting_system.dm [hunk 4]
```diff
@@ -94,70 +97,125 @@
...
 		changed = 1
 		SSlighting.changed_lights |= src
 
 //Remove current effect
 /datum/light_source/proc/remove_effect().
-	for(var/turf/T in effect)
-		T.update_lumcount(-effect[T])
+	for(var/turf/T in turfs_effect)
+		T.update_lumcount(-turfs_effect[T], turfs_direction[T])
+		T.affecting_lights -= src
```
Comment (tkdrg) [inline code/modules/lighting/lighting_system.dm]: I remember someone saying that `T.affecting_lights.Remove(src)` is faster, but I haven't tested it myself.

File: code/modules/lighting/lighting_system.dm [hunk 5]
```diff
@@ -94,70 +97,125 @@
...
 
-		if(T.affecting_lights && T.affecting_lights.len)
-			T.affecting_lights -= src
+		for(var/thing in RANGE_TURFS(1, T))
+			var/turf/neighbor = thing
```
Comment (tkdrg) [inline code/modules/lighting/lighting_system.dm]: Is this faster than `for(var/turf/neighbor in RANGE_TURFS(1, T))`?

Comment (RemieRichards) [inline code/modules/lighting/lighting_system.dm]: ```RANGE_TURFS()``` only returns turfs, so there's no point istyping each iteration of the loop (which is what a typed var in a for loop does, thus MrPerson simply typecasts it, "free" of charge.

this only works on lists where the data contained is of the same type or you're only acting on something available in their shared parent.

**Comment by YotaXP** [comment]
Try this, @MrPerson shadesmrp

File: code/modules/lighting/lighting_system.dm (hunk 4 above)
Comment (YotaXP) [inline code/modules/lighting/lighting_system.dm]: I think `+=`/`-=` are actually faster.

Comment (tkdrg) [inline code/modules/lighting/lighting_system.dm]: I'm sure that += is better yeah but I'm unsure about removing.

**Comment by Xhuis** [comment]
Seems like a poor imitation of Goon's lighting the way it is now. I'd prefer block lighting if this was the alternative.

**Comment by Incoming5643** [comment]
Reverse engineering takes time xhuis, have faith

**Comment by ike709** [comment]
I gotta agree with WJohn, it just looks way too blocky and not nearly as smooth as goon's. But if that's improved eventually, then :+1: x :100:

**Comment by optimumtact** [comment]
This will require everyone to have hardware rendering available correct?

File: code/modules/lighting/lighting_system.dm (hunk 1 above)
Comment (MrPerson) [inline code/modules/lighting/lighting_system.dm]: I did a test of this not too long back. It's the same either way.

File: code/modules/lighting/lighting_system.dm (hunk 5 above)
Comment (duncathan) [inline code/modules/lighting/lighting_system.dm]: Yeah, this is the fastest way to do a loop like this properly AFAIK

File: code/modules/lighting/lighting_system.dm (hunk 2 above)
Comment (MrPerson) [inline code/modules/lighting/lighting_system.dm]: This proc isn't the killer. The real problems are light_source/add_effect(), light_source/remove_effect(), turf/update_lumcount(), and turf/redraw_lighting().

File: code/modules/lighting/lighting_system.dm (hunk 4 above)
Comment (MrPerson) [inline code/modules/lighting/lighting_system.dm]: Actually what I really should be trying is probably ^, although remove_effect() is way cheaper than add_effect()

**Comment by MrPerson** [comment]
The blockiness is 100% the sprite. Same for the hardness of the split in the walls that WJohnson pointed out, although I'm not sure if that part is really ever going to be much better. Later on we can do occlusion on the walls to always make the back half one color though, which might help a bit.

This would be a LOT faster if turfs kept records of their neighbors. Keep that in mind. The slowest new part is easily having to call RANGE_TURFS() to grab nearby turfs.

@YotaXP 
Thanks. I'll try it out ASAP.

@optimumtact 
Yes they probably would, although I did intentionally choose a sprite/matrix combination that's at least almost playable in software mode. Basically the bottom left corner is correct but the other three corners are black. Software mode is weird.

**Comment by optimumtact** [comment]
This might look a little better if the fully black squares blended into the lit ones slightly, so the transition is not as abrupt

**Comment by MrPerson** [comment]
They do, it's just too abrupt. Worst-case I could give walls a different sprite.

@YotaXP 
Dunno if it's github or what but when I download that, it basically has no green in it and all the colors are really muted.

File: code/modules/lighting/lighting_system.dm (hunk 5 above)
Comment (RemieRichards) [inline code/modules/lighting/lighting_system.dm]: @duncathan yes it is, I am the authority here :P 
(for real though I've tested all possible loop formats byond has, I know which is best)

Comment (duncathan) [inline code/modules/lighting/lighting_system.dm]: I know @RemieRichards I was just here to reinforce your correctness!

**Comment by optimumtact** [comment]
turfs keeping a record of neighbours is probably fine, we're not exactly hitting our memory budget at all.

**Comment by tkdrg** [comment]
@optimumtact don't forget byond is 32-bit

**Comment by optimumtact** [comment]
I believe we are using around 500mb at the moment, which leaves us exactly 2.5g to play with

**Comment by Fox-McCloud** [comment]
suggestion: Add in colored lights.

Bay and Goon both have it--and it's really nice for ambient effects with things like candles and such.

**Comment by StormAGeddonz** [comment]
Gonna butt in here, there is already an ongoing project to make a plug and play module for this over at
https://github.com/PJB3005/GoonLight-Heist

http://i.imgur.com/ntfRnfG.png

**Comment by optimumtact** [comment]
M I T

**Comment by StormAGeddonz** [comment]
Indeed, its an MIT license meaning you can port it and redistribute it along with your code as agplv3

File: code/modules/lighting/lighting_system.dm (hunk 4 above)
Comment (PJB3005) [inline code/modules/lighting/lighting_system.dm]: Are you 100% sure about that because I know I tested it recently

**Comment by Bill-Luxe** [comment]
I dont know why, but i am still dream about this lighting on tg ._.

**Comment by PJB3005** [comment]
Goof said he's gonna port /vg/lighting after he's done with guns.

**Comment by Iamgoofball** [comment]
This is true

**Comment by Mervill** [comment]
ps gib

**Comment by PKPenguin321** [comment]
Pls goof

**Comment by Incoming5643** [comment]
I will fight you if you attempt it

**Comment by Iamgoofball** [comment]
We can optimize it

**Comment by Bawhoppen** [comment]
we should use tacolighting, it is the logical choice

**Pull request closed without merging**